classes/announcement/Collector.inc.php

```diff
@@ -0,0 +1,134 @@
+<?php
+/**
+ * @file classes/announcement/Collector.inc.php
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2000-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class announcement
+ *
+ * @brief A helper class to configure a Query Builder to get a collection of announcements
+ */
+
+namespace PKP\announcement;
+
+use HookRegistry;
+use Illuminate\Support\Facades\DB;
+use PKP\core\interfaces\CollectorInterface;
+
+class Collector implements CollectorInterface
+{
+    public DAO $dao;
+    public array $contextIds = [];
+    public string $searchPhrase = '';
+    public array $typeIds = [];
+    public int $count;
+    public int $offset;
+
+    public function __construct(DAO $dao)
+    {
+        $this->dao = $dao;
+    }
+
+    /**
+     * Filter announcements by one or more contexts
+     */
+    public function filterByContextIds(array $contextIds): self
+    {
+        $this->contextIds = $contextIds;
+        return $this;
+    }
+
+    /**
+     * Filter announcements by one or more announcement types
+     */
+    public function filterByTypeIds(array $typeIds): self
+    {
+        $this->typeIds = $typeIds;
+        return $this;
+    }
+
+    /**
+     * Filter announcements by those matching a search query
+     */
+    public function searchPhrase(string $phrase): self
+    {
+        $this->searchPhrase = $phrase;
+        return $this;
+    }
+
+    /**
+     * Limit the number of objects retrieved
+     */
+    public function limit(int $count): self
+    {
+        $this->count = $count;
+        return $this;
+    }
+
+    /**
+     * Offset the number of objects retrieved, for example to
+     * retrieve the second page of contents
+     */
+    public function offset(int $offset): self
+    {
+        $this->offset = $offset;
+        return $this;
+    }
+
+    /**
+     * @copydoc CollectorInterface::getQueryBuilder()
+     */
+    public function getQueryBuilder(): \Illuminate\Database\Query\Builder
+    {
+        $qb = DB::table($this->dao->table . ' as a');
+
+        if (!empty($this->contextIds)) {
```

Comment:
What do you think about adopting the convention where we use `NULL` to initialize the member on instantiation and check whether to apply a filter using `is_array` rather than `empty` here? I think calling code that passes in an empty array to `filterByContextIds` will expect to receive no results, rather than the condition being removed entirely.

Reply:
I like that. It will also prevent accidental misuse from exposing data for unwanted contexts and things.